### ISSUE TITLE
Measure Invariants

### ISSUE THREAD

**Issue by ranjitjhala** [ucsd-progsys/liquidhaskell#1198]
@nikivazou why does the below program fail 

```haskell
module Term where

{-@ data Tree [sz] @-}
data Tree = Leaf | Node Tree Tree

{-@ measure sz @-}
{-@ sz :: Tree -> Nat @-}
sz :: Tree -> Int
sz Leaf = 0
sz (Node t1 t2) = 1 + sz t1 + sz t2
```

**FAIL** with the odd (termination) error: 

```
/Users/rjhala/research/sf/liquid-sf/term.hs:10:23-27: Error: Liquid Type Mismatch

 10 | sz (Node t1 t2) = 1 + sz t1 + sz t2
                            ^^^^^

   Inferred type
     VV : {v : Tree | (Term.sz v < Term.sz ?a => Term.sz v >= 0)
                      && v == t1}

   not a subtype of Required type
     VV : {VV : Tree | Term.sz VV < Term.sz ?a

                       && Term.sz VV >= 0}

   In Context
     t1 : {t1 : Tree | Term.sz t1 < Term.sz ?a => Term.sz t1 >= 0}

     ?a : {?a : Tree | Term.sz ?a < Term.sz ?a => Term.sz ?a >= 0}
```

Where is the antecedent `Term.sz t1 < Term.sz ?a => ...` that guards `Term.sz t1 >= 0` coming from? 

Why is it necessary? 

Note that the code passes if you add:

```haskell
{-@ invariant {v:Tree | 0 <= sz v } @-}
```

but this is unnecessary, as it should be generated from the measure? What am I missing here?

**Comment by nikivazou** [comment]
I run into this again.... 

The guard is necessary, I believe for soundness: it says that the result refinement (that `0 <= sz t`) is valid only on `t` that is provably smaller than the argument of the function. 

Should I remove it? 
I believe it is required.....

In other words... I am not sure how to prove termination of the "termination metric" function.
And I suspect we need some notion of syntactic termination checking for those...

**Comment by ranjitjhala** [comment]
The `<` is not useful in the above because the ordering 
is undefined; i.e. the refinement logic can do nothing with 
it other than conservatively assume that it does not hold.

Instead we *definitely* need to implement the standard 
`inductive-datatype` definition, doing a syntactic check 
e.g. where the following _is inductive_,

```haskell
data List a = Emp | Cons a (List a)
```

but where the following is _not_:

```haskell
data List a = Cons a (List a)
```

Once this is done, a `structural-termination` 
check is straightforward (syntactic), and then
we should check that all measures satisfy 
`structural-termination`, and that's it.

Both the above should be easy to do and the 
above error is really puzzling, so we may as 
well do it...?

**Comment by nikivazou** [comment]
We do not need to implement structural-termination **for all** measures. 
Just for the measures we define as "sizes" for the ADTs. Then we can use these sizes to prove termination, semantically for anything else.... 


You are correct, may as well do it :) Will give it a try because it is really annoying...

**Comment by ranjitjhala** [comment]
To be clear: I think the `structural-termination` check is simply, 

*if* there is recursion, then 

1. The underlying datatype is `inductive`, and
2. The the parameter is a constituent of something smaller...

**Comment by ranjitjhala** [comment]
(Yes, good point re: only needing this for the `size` measures...!)

**Comment by nikivazou** [comment]
Fixed with #1244

**Comment by nomeata** [comment]
There is still a test case for this in the `neg` folder:
```
$ cat ./tests/neg/T1198.2.hs
module Term where

{-@ data Tree [sz] @-}
data Tree a =  Node (Tree a) (Tree a)

{-@ measure sz @-}
sz :: Tree a -> Int
sz (Node t1 t2) = 1 + sz  t1 + sz  t2
```
It passes with https://github.com/ucsd-progsys/liquidhaskell/pull/1278 however. That’s good, right?
